Fixes #10134 - Stdout/err in Rails logger now and better CR logging

lzap opened the pull request:
This is just an idea. When running with Rails DEBUG mode on, we will see
requests/replies from compute resources.

Those fog drivers which leverage `excon` library accepts `connection_options`.
Special ones are libvirt (does not currently accept that) and vmware (it has
it's own debug capabilities - to be merged:
https://github.com/fog/fog/pull/3530)

lzap committed:
Fixes #10134 - Stdout/err in Rails logger and better CR logging

lzap commented:
The vmware patch was merged into fog master, when testing vmware make sure you have master. Note not all drivers do not (yet) support connection_options. One will likely never do (libvirt). I can remove that line perhaps.

lzap commented on config/initializers/1_redirect_stdout.rb:
```diff
@@ -0,0 +1,6 @@
+# Rails log.
+
+$stderr.reopen Rails.logger.instance_variable_get(:@logdev).dev
+$stdout.reopen Rails.logger.instance_variable_get(:@logdev).dev
```
Ok added this. Might be dangerous if someone closes (or reopens) the device. Alternative might be an IO wrapper: http://stackoverflow.com/questions/9637092/redirect-stderr-to-logger-instance

domcleal commented on config/initializers/1_redirect_stdout.rb:
```diff
@@ -0,0 +1,6 @@
+# Using this technique it is possible to redirect all the output to standard
+# Rails log.
+
+$stderr.reopen Rails.logger.instance_variable_get(:@logdev).dev
```
Tests are failing here:

<pre>
00:03:40.685 rake aborted!
00:03:40.685 NoMethodError: undefined method `dev' for nil:NilClass
00:03:40.686 /var/lib/workspace/workspace/test_develop_pr_core@2/database/postgresql/ruby/1.9.3/config/initializers/1_redirect_stdout.rb:5:in `<top (required)>'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:245:in `load'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:245:in `block in load'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:236:in `load_dependency'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:245:in `load'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/engine.rb:593:in `block (2 levels) in <class:Engine>'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/engine.rb:592:in `each'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/engine.rb:592:in `block in <class:Engine>'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/initializable.rb:30:in `instance_exec'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/initializable.rb:30:in `run'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/initializable.rb:55:in `block in run_initializers'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/initializable.rb:54:in `each'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/initializable.rb:54:in `run_initializers'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/application.rb:136:in `initialize!'
00:03:40.686 /usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pr_core-2/gems/railties-3.2.21/lib/rails/railtie/configurable.rb:30:in `method_missing'
00:03:40.686 /var/lib/workspace/workspace/test_develop_pr_core@2/database/postgresql/ruby/1.9.3/config/environment.rb:12:in `<top (required)>'
</pre>

domcleal commented:
When rebased, this should probably register a new logger and use that if enabled.

lzap commented:
Will do and fix the tests issue. Do you like the idea?

lzap commented:
Or rather - my implementation :)

domcleal commented:
I like the idea, but will decide if I like the impl when it's in a state to be reviewed!

lzap commented:
Ok rebased on top of our new logging. Finished the instrumentator. Added new appender called `crs` (compute resources) where it goes. Warning: VmWare has it's own logging mechanism and it log to STDOUT.

I dropped the STDOUT/STDERR redirection, because this does not work and cause problems (app won't boot, cannot use ruby debugger on development mode). The only drawback is if user turns on debug for the `crs` logger, then messages will be actually logged to stdout - from my experience, this does not end up in httpd logs, but it is thrown away.

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+module FogExtensions
```
This isn't a concern, I'd suggest putting it into app/services/foreman/

domcleal replied:
Scrap that, if you take my AS notifications advice, since it would probably then go into an initialiser.

domcleal commented on config/application.rb:
```diff
@@ -144,7 +144,11 @@ class Application < Rails::Application
+    Foreman::Logging.add_loggers(SETTINGS[:loggers] || {
+      :app => {:enabled => true},
+      :sql => {:enabled => true},
+      :crs => {:enabled => true}
```
please also add this to settings.yaml.example

trailing comma too

domcleal replied:
Should we default this to off, since it may log a lot?  It's easy to enable now from settings.yaml if needed.

domcleal replied:
This reminded me to move the LDAP stuff to a separate logger too, in #2437, which I switched off by default and improved enabling a single logger from settings.yaml.

domcleal replied:
I'd suggest calling it 'compute' too, so it's clearer.

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+                logger.debug("#{params.inspect}")
+              end
+            rescue => e
+              logger.debug("Error during fog debug instrumentation: #{e}")
```
backtrace too

via Foreman::Logging.exception

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+                logger.debug("Date: #{params[:headers]['Date']}\n")
+                logger.debug("Body: #{params[:body]}")
+              elsif name.include?('.retry')
+                logger.debug("#{params.inspect}")
```
no need to interpolate, ditto on 53

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+                  end
+                end
+                logger.debug("#{params[:method]} #{params[:path]}#{query} HTTP/1.1" )
+                logger.debug("User-Agent: #{params[:headers]['User-Agent']}")
```
Why are you printing only some of params[:headers] and not the whole hash?

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+                tmp_query = ''
+                if params.key?(:query) && !params[:query].nil?
+                  params[:query].each do |key, value|
+                    tmp_query += "#{key}=#{value}&"
```
`params[:query].map { |p| "#{p[0]}=#{p[1]}" }.join("&")` saves the chomp and explicit string building.

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+                logger.debug("Content-Type: #{params[:headers]['Content-Type']}")
+                logger.debug("Date: #{params[:headers]['Date']}")
+                params[:headers].each do |key, value|
+                  if !['Content-Length', 'Content-Type', 'Date'].include?(key)
```
Why are you printing some headers explicitly and others not?

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+                    logger.debug("#{key}: #{value}")
+                  end
+                end
+                logger.debug("Date: #{params[:headers]['Date']}\n")
```
dupe of 42

domcleal commented on app/models/concerns/fog_extensions/debug/debug_http_instrumentor.rb:
```diff
@@ -0,0 +1,67 @@
+      class << self
+        attr_accessor :events
+
+        def instrument(name, params = {}, &block)
```
Rather than writing something against the instrumentation API itself, it'd might be preferable to pass in `ActiveSupport::Notifications` and then write against `ActiveSupport::LogSubscriber`.

See config/initializers/ldap_instrumentation.rb for an example.  Just add another initialiser that attaches to the names of the events we're interested in?

domcleal commented:
It's a rather complex piece of logging code, I'd be much happier if you could add a small test that sends a few known event fixtures through it and ensures it doesn't throw any exceptions (or catch them, since it does rescue), as it could be fragile.

Also note the rubocop failures.

domcleal commented:
Also, the commit message still talks about stdout/err.

domcleal commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -103,7 +103,12 @@ def self.provider_friendly_name
+        :google_project      => project,
+        :google_client_email => email,
+        :google_key_location => key_path,
+        :instrumentor        => FogExtensions::Debug::DebugHttpInstrumentor)
```
This may result in double-logging, since google-api-client already opportunistically uses Rails.logger if available at debug level.  We might want to simply set Google::APIClient.logger in an initialiser.

tbrisker commented:
@lzap ,
Thank you very much for your efforts!
This PR has been open for a very long time with no activity.
Please indicate if you plan to continue working on it or not; if not, feel free to close it.
If you do not respond either way within a couple of weeks, I will assume this PR is abandoned and will go ahead and close it for you.

tbrisker commented:
Dear @lzap,
It has been 20 days since my last comment with no replies. 
I am assuming this pull request has been abandoned and closing it.
If you wish to continue working on this in the future, feel free to reopen.
On behalf of the entire Foreman team, I wish to thank you for your efforts here and hope to see more contributions from you in the future.
